Add DQ ranking statistic to coinc search workflow

dethodav opened the pull request:
This pull request is still work in progress as it depends on #3804 and includes commits that will be added as part of that PR.

This PR adds support for the new DQ ranking statistic to be used in the coinc search workflow. This required a number of changes to `pycbc_make_coinc_search_workflow` to support the new executables and workflow structure. However, this new code should be fully back-compatible.

titodalcanton commented:
@dethodav as we are close to merging #3804, please make sure this is appropriately rebased. At the moment it conflicts with master and appears to introduce many of the things already introduced in #3804.

titodalcanton commented:
What is the status of this? Is it still relevant?

titodalcanton commented:
@maxtrevor you may actually be able to push to Derek's branch. Give it a try, if it does not work you can create a new branch, though I would prefer to continue to use this PR for cleanliness.

dethodav commented:
@titodalcanton @maxtrevor I can take a look at this again - apologies for letting this sit for so long

dethodav commented:
@titodalcanton @maxtrevor I've gone ahead and rebased to the current master branch and implement some of the previous suggestions by @titodalcanton. I'm currently rerunning a full workflow to confirm that everything is still working after the rebase.

maxtrevor commented:
@titodalcanton @dethodav whats next to keep this moving?

titodalcanton commented on pycbc/workflow/dq.py:
```diff
@@ -0,0 +1,203 @@
+        assert len(ifo_insp)==1
+        ifo_insp = ifo_insp[0]
+        flag_name = flag_str
+        logging.info("Creating job for flag %s" % (flag_name))
```
```suggestion
        logging.info("Creating job for flag %s", flag_name)
```

titodalcanton commented on bin/all_sky_search/pycbc_combine_statmap:
```diff
@@ -57,9 +57,12 @@ for fi in files:
-# Convert segmentlistdict to a list ('seglists') of segmentlists
-# then numpy.sum(seglists, axis=0) does seglists[0] + seglists[1] + ...
-foreground_segs = numpy.sum(list(indiv_segs.values()), axis=0)
+if len(indiv_segs.values())==1:
```
```suggestion
if len(indiv_segs.values()) == 1:
```

titodalcanton commented on bin/plotting/pycbc_plot_bank_bins:
```diff
@@ -91,12 +97,12 @@ parser.add_argument('--log-y', action='store_true',
 
 bank = H5BankFile(args.bank_file, 'r')
-bank.f_lower = args.f_lower
+f_lower  = args.f_lower or bank['f_lower'][:]
```
```suggestion
f_lower = args.f_lower or bank['f_lower'][:]
```

titodalcanton commented on bin/plotting/pycbc_plot_dq_likelihood_vs_time:
```diff
@@ -0,0 +1,86 @@
+parser = argparse.ArgumentParser(description=__doc__)
+parser.add_argument('--version', action='version', version=version)
+parser.add_argument('--verbose', action="store_true")
+parser.add_argument("--ifo", type=str,required=True)
```
```suggestion
parser.add_argument("--ifo", type=str, required=True)
```

titodalcanton commented on bin/plotting/pycbc_plot_dq_likelihood_vs_time:
```diff
@@ -0,0 +1,86 @@
+
+bin_name = args.background_bin
+
+if bin_name not in f['%s/dq_vals'%ifo].keys():
```
```suggestion
if bin_name not in f[f'{ifo}/dq_vals'].keys():
```

titodalcanton commented on bin/plotting/pycbc_plot_dq_likelihood_vs_time:
```diff
@@ -0,0 +1,86 @@
+
+if bin_name not in f['%s/dq_vals'%ifo].keys():
+    raise ValueError("Background bin name not found in DQ file")
+else:
+    yvals = f['%s/dq_vals/%s'%(ifo,bin_name)][:]
```
No need for the `else` block here, the code just carries on.
```suggestion
yvals = f[f'{ifo}/dq_vals/{bin_name}'][:]
```

titodalcanton commented on bin/plotting/pycbc_plot_dq_likelihood_vs_time:
```diff
@@ -0,0 +1,86 @@
+else:
+    yvals = f['%s/dq_vals/%s'%(ifo,bin_name)][:]
+
+xvals = f['%s/times'%ifo][:]
```
```suggestion
xvals = f[f'{ifo}/times'][:]
```

titodalcanton commented on bin/workflows/pycbc_make_coinc_search_workflow:
```diff
@@ -494,6 +530,26 @@ for key in final_bg_files:
+
+# DQ log likelihood plots
+for dqf,dql in zip(dqfiles,dqfile_labels):
+    outdir = rdir['data_quality/%s_trigger_rates_%s' % (dqf.ifo,dql)]
```
```suggestion
    outdir = rdir[f'data_quality/{dqf.ifo}_trigger_rates_{dql}']
```

titodalcanton commented on bin/workflows/pycbc_make_coinc_search_workflow:
```diff
@@ -494,6 +530,26 @@ for key in final_bg_files:
+    outdir = rdir['data_quality/%s_trigger_rates_%s' % (dqf.ifo,dql)]
+    wf.make_dq_trigger_rate_plot(workflow, [dqf], outdir, tags=[dql,'full_data'])
+
+    outdir = rdir['data_quality/%s_percentiles_%s' % (dqf.ifo,dql)]
```
```suggestion
    outdir = rdir[f'data_quality/{dqf.ifo}_percentiles_{dql}']
```

titodalcanton commented on pycbc/workflow/dq.py:
```diff
@@ -0,0 +1,203 @@
+        return node
+
+class PyCBCCalculateDQFlagExecutable(Executable):
+    # current_retention_level = Executable.ALL_TRIGGERS
```
Any reason not to remove this line?

titodalcanton commented:
I just left a bunch of minor style suggestions. As a more general comment, the two plotting scripts added by this PR look very similar, does it make sense to consolidate them into a single script?

@dethodav was also running one more test to check everything is ok, I would like to hear from that before approving.

dethodav commented:
@titodalcanton My rerun of the workflow after rebasing was successful: 

https://ldas-jobs.ligo.caltech.edu/~derek.davis/cbc/preO4/pycbc_idq/merge_request_tests/dq-workflow-simple-gwosc-chunkgwosc-1/

I did identify an issue where one of the plotting scripts can fail when small amounts of data are analyzed. I fixed this separately in this PR: https://github.com/gwastro/pycbc/pull/4153

I'll work on addressing these comments and other style suggestions as soon as I can.

maxtrevor commented:
@dethodav Any updates on this?

titodalcanton commented:
Most of my comments above can simply be accepted by clicking the "commit suggestion" button, so should be fairly quick to address.

maxtrevor commented:
@dethodav any progress on this? We really need to get this merged in ASAP

titodalcanton commented:
I took the liberty of applying most of my suggestions, as they were really trivial. I think a rebase on master is needed now to make the CI tests pass.

dethodav commented:
@titodalcanton @maxtrevor I've finally updated they PR - let me know if there's any other changes I should make!

titodalcanton reviewed (approved):
Thanks @dethodav @maxtrevor. I think it is time to merge this.
